apps/coordinator/ts/scheduler/scheduler.service.ts

```diff
@@ -0,0 +1,282 @@
+import { SchedulerRegistry } from "@nestjs/schedule";
+
+import { ErrorCodes } from "../common";
+import { RedisService } from "../redis/redis.service";
+import { IIdentityScheduledPoll, IScheduledPoll } from "../redis/types";
+import { getPollKeyFromObject } from "../redis/utils";
+import { SessionKeysService } from "../sessionKeys/sessionKeys.service";
+
+import {
+  IIsPollScheduledResponse,
+  IPollEndDateAndIsTalliedResponse,
+  IIdentityPollWithSignerArgs,
+  ISchedulePollWithSignerArgs,
+} from "./types";
+
+@Injectable()
+export class SchedulerService implements OnModuleInit {
+  /**
+   * Logger
+   */
+  private readonly logger: Logger;
+
+  /**
+   * Create a new scheduler service instance.
+   * @param sessionKeysService - session keys service to manage session keys
+   * @param redisService - redis service to interact with Redis database
+   * @param schedulerRegistry - scheduler registry to manage scheduled tasks
+   */
+  constructor(
+    private readonly sessionKeysService: SessionKeysService,
+    private readonly redisService: RedisService,
+    private readonly schedulerRegistry: SchedulerRegistry,
+  ) {
+    this.logger = new Logger(SchedulerService.name);
+  }
+
+  /**
+   * Instantiate and schedule the saved polls in the database
+   * @dev it is a lifecycle hook provided by NestJS that is called after all dependencies
+   * are injected and the module is fully initialized.
+   * This ensures your service is ready and all dependencies are available.
+   */
+  async onModuleInit(): Promise<void> {
+    await this.restoreTimeouts();
+  }
+
+  /**
+   * Check if poll is scheduled for finalization
+   * @param args - Arguments for checking poll finalization
+   * @returns true if poll is scheduled for finalization, false otherwise
+   */
+  async isPollScheduled(scheduledPoll: IIdentityScheduledPoll): Promise<IIsPollScheduledResponse> {
+    const key = getPollKeyFromObject(scheduledPoll);
+    const { maciAddress, pollId, chain } = scheduledPoll;
+
+    const storedPoll = await this.redisService.get(key);
+    if (!storedPoll) {
+      return { isScheduled: false };
+    }
+
+    try {
+      this.schedulerRegistry.getTimeout(key);
+    } catch (error) {
+      // delete redis data so user has to reschedule again
+      await this.redisService.delete(key);
+      return { isScheduled: false };
+    }
+
+    const { isPollTallied } = await this.pollEndDateAndIsTallied({
+      maciAddress,
+      pollId,
+      chain,
+    });
+
+    if (isPollTallied) {
+      this.deleteScheduledPoll({ maciAddress, pollId, chain });
+      this.logger.error(`Error: ${ErrorCodes.POLL_ALREADY_TALLIED}, poll has already been finalized`);
+      throw new Error(ErrorCodes.POLL_ALREADY_TALLIED.toString());
+    }
+
+    return { isScheduled: true };
+  }
+
+  /**
+   * Return the poll end date and whether it is tallied
+   * @param args - Arguments to retrieve poll data
+   */
+  async pollEndDateAndIsTallied({
```

Comment:
Let's rename it. It should be a verb like `getPollFinalizationData` or something like this and let's add return type to docs.